### PR TITLE
Adds `r2.OptMaxRedirects` which overrides the default of 10

### PR DESCRIPTION
**Pull request by wcharczuk**
Adds an option to set the max redirects, but also adds a utility to check if the error returned on the request was the result of the maximum number of redirects being reached.

### CODE REVIEW

File: r2/opt_max_redirects.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package r2
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/blend/go-sdk/logger"
+
```
Comment (dhermes) [inline r2/opt_max_redirects.go]: Empty line not needed?

File: r2/opt_max_redirects.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/blend/go-sdk/ex"
+	"github.com/blend/go-sdk/logger"
+)
+
+// OptMaxRedirects tells the http client to only follow a given
+// set of redirects, overriding the standard library default of 10.
```
Comment (dhermes) [inline r2/opt_max_redirects.go]: "set of redirects" --> number

File: r2/opt_max_redirects.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/blend/go-sdk/ex"
+	"github.com/blend/go-sdk/logger"
+)
+
+// OptMaxRedirects tells the http client to only follow a given
+// number of redirects, overriding the standard library default of 10.
+// If a maximum number of redirects is reached, an exception of class `http.ErrUseLastResponse`
+// will be returned, with the redirect history as the exception message.
+// NOTE: that will make the exception message incredibly large.
```
Comment (dhermes) [inline r2/opt_max_redirects.go]: "will make" -> "may make"?

File: r2/opt_max_redirects.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+func OptMaxRedirects(maxRedirects int) Option {
+	return func(r *Request) error {
+		if r.Client == nil {
+			r.Client = &http.Client{}
+		}
+		r.Client.CheckRedirect = func(r *http.Request, via []*http.Request) error {
+			if len(via) >= maxRedirects {
+				r = r.WithContext(logger.WithAnnotations(r.Context(), logger.Annotations{
+					"via": urlStrings(via),
+				}))
```
Comment (dhermes) [inline r2/opt_max_redirects.go]: How does this help? You'd need to set `r.Context` directly to leave this scope? Or does `WithContext` modify in place? (If yes, no reason to assign to `r`)

File: r2/opt_max_redirects.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+			}
+			return nil
+		}
+		return nil
+	}
+}
+
+// ErrIsTooManyRedirects returns if the error is too many redirects.
+func ErrIsTooManyRedirects(err error) bool {
+	if typed, ok := err.(*url.Error); ok {
```
Comment (dhermes) [inline r2/opt_max_redirects.go]: Why the `url.Error` check?

File: r2/opt_max_redirects.go (hunk 4 above)
Comment (wcharczuk) [reply r2/opt_max_redirects.go]: with context returns the modified r, context is set for you by net/http

File: r2/opt_max_redirects.go (hunk 5 above)
Comment (wcharczuk) [reply r2/opt_max_redirects.go]: because the go authors are mean

Comment (dhermes) [reply r2/opt_max_redirects.go]: I'm not sure I follow. When would `ErrIsTooManyRedirects()` be used?

File: r2/opt_max_redirects.go (hunk 4 above)
Comment (dhermes) [reply r2/opt_max_redirects.go]: Gotcha: https://github.com/golang/go/blob/go1.13.7/src/net/http/request.go#L360

File: r2/opt_max_redirects.go (hunk 5 above)
Comment (wcharczuk) [reply r2/opt_max_redirects.go]: see the test

File: r2/opt_max_redirects_test.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		http.Redirect(rw, r, pingURL, http.StatusTemporaryRedirect)
+		return
+	}))
+	defer pong.Close()
+
+	pingURL = ping.URL
+	pongURL = pong.URL
+
+	_, err := New(pingURL, OptMaxRedirects(32)).Discard()
+	assert.True(ErrIsTooManyRedirects(err))
```
Comment (wcharczuk) [inline r2/opt_max_redirects_test.go]: @dhermes usage

Comment (dhermes) [reply r2/opt_max_redirects_test.go]: DERP 🤦‍♂

**Pull request merged**